tensor product of `GeneralizedStabilizer`

Fe-r-oz opened the pull request:
Building on #259, Basic API as stated in #260 are introduced:

- [x]  sm⊗sm

Checklist: 

- [x] The code is properly formatted and commented.
- [x] Substantial new functionality is documented within the docs.
- [x] All new functionality is tested.
- [x] All of the automated tests on github pass.

Krastanov commented:
This would be pretty difficult to review as is. Here are a few things that might help (do not worry about it too much if it is the first time you are experiencing this workflow, it takes a while to get accustomed to it):

- If the pull request is about finishing #259, it should be a pull request for merging into the branch noncliff, not into the master branch like this one. In that case it would also not have the need for the complicated `Merge branch 'master' into nonstab` step.
- If the pull request is about adding extra interfaces as described in #260, then it probably should be done independently, not on top of #259 (most easily, that would be **after** #259 is finished)

I personally would avoid trying to merge the #260 API improvements (what this PR seems to be) until #259 itself is finished (it currently has a buggy implementation).

Fe-r-oz commented:
Pardon for the inconvenience. I hope that https://github.com/QuantumSavory/QuantumClifford.jl/pull/345/commits/46bfaacf3ef37277a255d9a6e54bbda7a8256a3a makes review easier than before. 

P.S. When I was resolving merge conflicts (that lead to first commit, weirdly I added test for `sm ⊗ sm` test when  resolving merge conflict in `test_stab.jl.` It was not visible in the second commit which is going to get reviwed. So, I modified it so it's visible now.  Also, JET throws error which can be resolved increasing to `>=29`

Fe-r-oz commented:
I understand.  Actually, I was unable to push commits because of merge conflicts that's why the merge branch master into noncliff was needed. Otherwise, it's show 'Resolve merge conflicts'. Actually, I am somewhat confused about this workflow, but I am learning.

I think these tasks from #260 are independent of the `expect`. Doctests for each of these tasks have been added to demonstrate the result. 

I think the bug is related to `expect`, and to resolve it, I implemented the Algorithm 2 from Ted paper as well which is show below. But this did not resolve the expect bug even though.

Hence, I moved towards implementing the tasks from #260 as they were independent of the bug which I think is coming from `expect` as by starting with smaller steps.

My original thinking was that I should divide it into four phases:
phase 0: resolve some basic todos e.g in `apply! `where `filter! `can be used instead of `delete `
phase 1 & 2: complete the functionality that is indepentent of that bug coming from `expect` by dividing the tasks from #260 into two parts
phase 3: implement the exact algorithms (1, 2) from Ted paper to resolve the bug. This way, I would save time and also be able to do fruitful work in forward direction.

I tried to implement algorithm 2 as closely as possible from Ted's paper .  I will focus on the algorithms from Ted's paper.

Algorithm 2
```
function expect(p::PauliOperator, s::GeneralizedStabilizer)
    dict = s.destabweights
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    tone = one(dtype)
    newdict = typeof(dict)(tzero)
    phase, b, c = rowdecompose(p, s.stab)
    if all(x -> x == 0, b)
        for ((dᵢ, dⱼ), χ) in dict
            if ((-tone)^(dot(dᵢ, c)))*im^phase == 1 && ((-tone)^(dot(dⱼ, c)))*im^phase == 1
                newdict[(dᵢ, dⱼ)] = get(newdict, (dᵢ, dⱼ), 0) + χ
            end
        end
        s.destabweights = newdict
        return s
    else
        k = _create_k(b)
        for ((dᵢ, dⱼ), χ) in dict
            q = 1
            dᵢ′ = copy(dᵢ)
            dⱼ′ = copy(dⱼ)
            if dot(dᵢ, k) == 1
                q *= ((-tone)^(dot(dᵢ, c)))*im^phase
                dᵢ′ = b .⊻ dᵢ
            end
            if dot(dⱼ, k) == 1
                q *= (-tone)^(dot(dⱼ, c))*conj(im^phase)
                dⱼ′ = b .⊻ dⱼ
            end
            newdict[(dᵢ′, dⱼ′)] = get(newdict, (dᵢ′,dⱼ′), 0) + χ * q / 2
        end
         s.destabweights = newdict
        return s
    end
end

function _create_k(b)
    k = zeros(Int, length(b))
    idx = findfirst(x -> x == 1, b)
    if idx !== nothing
        k[idx] = true
    end
    return k
end
```

Algorithm 1 (About conjugating destabs)
```
function conj_destabs!(state::GeneralizedStabilizer, gate::AbstractCliffordOperator)
    dict = state.destabweights
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    newdict = typeof(dict)(tzero)
    gate_fields = fieldnames(typeof(gate))
    g_field = nothing
    for field in gate_fields
        if getfield(gate, field) isa Vector{UInt64}
            g_field = field
            break
        end
    end
    if g_field === nothing
        error("No Vector{UInt64} field found in the gate object")
    end
    g_part = getfield(gate, g_field)
    C = Vector{Bool}(g_part)
    for ((dᵢ, dⱼ), χ) in dict
        # Compute dᵢ' and dⱼ'
        dᵢ′ = C .* dᵢ .* C
        dⱼ′ = C .* dⱼ .* C
        newdict[(dᵢ′, dⱼ′)] = χ
    end
    state.destabweights = newdict
    state
end
```

Fe-r-oz commented:
Please help review this PR, @Krastanov! Thank you very much!!

P.S. changelog: tensor product of `GeneralizedStabilizer`

Krastanov commented:
I moved the test that was added to a more appropriate file.

I also added a new test. The existing test was simply checking that errors are not raised, but it was not checking that the operation is actually correct. The new test provides that additional verification by converting to `QuantumOptics.Operator` and using its `tensor` method.

Fe-r-oz commented:
> I also added a new test. The existing test was simply checking that errors are not raised, but it was not checking that the operation is actually correct. The new test provides that additional verification by converting to `QuantumOptics.Operator` and using its `tensor` method.

Thank you so very much for this much needed and very cool test!! 🙏🏼 🙏🏼 ❤️ Indeed, the existing test had several shortcomings because it didn't test the actual correctness, just whether conditions were met. This new test is amazing!!

Fe-r-oz commented:
Thank you very much for your suggestions!! Please help review this PR, Thank you!!

The CI documentation error was likely caused due to noncliff paper `reference.bib` issue based on the stacktrace in this and previous PRs, so that is resolved as well. The error was because `@proceedings` require `booktitle` key as per CI documentation but the citation `@proceedings` didn't have the `booktitle` key. 

This bibtex is from the Semantic Scholar website: [link
]( https://www.semanticscholar.org/paper/A-generalization-of-the-stabilizer-formalism-for-Yoder/b200efe1709d07ffc1b5b7bd90e61c09e2729bdf)

```
@inproceedings{Yoder2012AGO,
  title={A generalization of the stabilizer formalism for simulating arbitrary quantum circuits},
  author={Theodore J. Yoder},
  year={2012},
  url={https://api.semanticscholar.org/CorpusID:18406560}
}
```

The bibtex available used in Semantic Scholar uses `@proceedings` which requires `booktitle` key according to CI documentation. So, I used the `howpublished` key for url stuff and used `@misc` instead of `@proceedings` to resolve the CI documentation error.
